src/GovUk.Education.ExploreEducationStatistics.Content.Model/Predicates/ReleaseVersionPredicates.cs

```diff
@@ -47,4 +47,26 @@ public static class ReleaseVersionPredicates
                 maxVersion => maxVersion,
                 (releaseVersion, _) => releaseVersion);
     }
+
+    /// <summary>
+    /// Filters a sequence of <see cref="IQueryable{T}"/> of type <see cref="ReleaseVersion"/> to only include the latest
+    /// published versions of each release.
+    /// </summary>
+    /// <param name="releaseVersionsQueryable">The source <see cref="IQueryable{T}"/> of type <see cref="ReleaseVersion"/> to filter.</param>
+    /// <param name="releaseId">Unique identifier of a release to filter by.</param>
+    /// <param name="publishedOnly">Flag to only include published release versions.</param>
+    /// <returns>An <see cref="IQueryable{T}"/> of type <see cref="ReleaseVersion"/> that contains elements from the input
+    /// sequence filtered to only include the latest versions of each release.</returns>
+    public static IQueryable<ReleaseVersion?> LatestReleaseVersion(this IQueryable<ReleaseVersion> releaseVersionsQueryable,
+        Guid releaseId,
+        bool publishedOnly = false)
+    {
+        return releaseVersionsQueryable
+            .Where(releaseVersion => releaseVersion.ReleaseId == releaseId)
+            .Where(releaseVersion => releaseVersion.Version == releaseVersionsQueryable
+                .Where(latestVersion => latestVersion.ReleaseId == releaseId)
+                .Where(latestVersion => !publishedOnly || latestVersion.Published.HasValue)
+                .Select(latestVersion => (int?)latestVersion.Version)
+                .Max());
```

Comment:
I've not seen this nesting before, I'm assuming this is working in both EF core in-memory  db for unit tests and for real with SQLServer?

Do you know what the difference is in the generated query with this nesting vs. if this was done as a join?

Reply:
I believe we've used subqueries like this elsewhere. Tough to find documentation for it though: https://visualstudiomagazine.com/articles/2018/10/01/subqueries-in-linq.aspx

I double checked that the generated query was nested:
```
dbug: Microsoft.EntityFrameworkCore.Database.Command[20100]
      Executing DbCommand [Parameters=[@__releaseId_0='bf3167d4-6a81-4b5e-bab7-08dc554c3456'], CommandType='Text', CommandTimeout='30']
      SELECT TOP(2) [r].[Id], [r].[ApprovalStatus], [r].[Created], [r].[CreatedById], [r].[DataGuidance], [r].[NextReleaseDate], [r].[NotifiedOn], [r].[NotifySubscribers], [r].[PreReleaseAccessList], [r].[PreviousVersionId], [r].[PublicationId], [r].[PublishScheduled], [r].[Published], [r].[RelatedInformation], [r].[ReleaseId], [r].[ReleaseName], [r].[Slug], [r].[SoftDeleted], [r].[TimePeriodCoverage], [r].[Type], [r].[UpdatePublishedDate], [r].[Version]
      FROM [ReleaseVersions] AS [r]
      WHERE [r].[SoftDeleted] = CAST(0 AS bit) AND [r].[ReleaseId] = @__releaseId_0 AND [r].[Version] = (
          SELECT MAX([r0].[Version])
          FROM [ReleaseVersions] AS [r0]
          WHERE [r0].[SoftDeleted] = CAST(0 AS bit) AND [r0].[ReleaseId] = @__releaseId_0 AND [r0].[Published] IS NOT NULL)
```

Interestingly, in the article I linked above from 2018, it seemed while you could do subqueries in linq at that time, the article suggests at the end that it may have been generating multiple queries behind the scenes. Seems they've actually generated nested queries now.

Reply:
Cool, I'm not sure what the difference in performance between the join and the nested query would be, but this seems fine.

Just thinking about this reminded me that as an alternative to adding this new method, calls to this should have also been possible by just making this instead I think? 🙂 

```
_contentDbContext.ReleaseVersions
.LatestReleaseVersions()
.FirstOrDefaultAsync(version => version.ReleaseId == releaseId)
```

Not saying you switch to this but if you have any time it would be interesting to see what this would have generated. As it's all in SQL rather than in-memory it should still only fetch at most one result back from the db.